### ISSUE TITLE
Check for source formatting / whitespace problems

### ISSUE THREAD

**Issue by n1tehawk** [mpeterv/luacheck#79]
To help with keeping a consistent coding style, it might be useful if LuaCheck would point out whitespace issues. I'm thinking along the lines of 
```lua
{
["611"] = "line contains only whitespace",
["612"] = "trailing whitespace on code line",
["613"] = "trailing whitespace on non-code line",
["621"] = "inconsistent indentation (SPACE followed by TAB)",
}
```
If this is considered useful, I could contribute some proof-of-concept code.

Regards, NiteHawk

**Comment by mpeterv** [comment]
Yes, this would be useful!

Ideally luacheck would also check for correct amount of whitespace around operators, parens, etc., but that would require more cooperation with the parser, and also it will be more complex to support customization. The warnings you suggested don't seem to need any options.

I haven't touched luacheck code in a while, and contributions are most welcome! I think that the way to do this would be to modify the lexer to consider whitespace a token, with exactly one whitespace token between each two consecutive normal tokens. Then the parser can call a function provided by the checker on each token. That function would check whitespace tokens and issue warnings.

**Comment by n1tehawk** [comment]
Thanks for diligently following my initial commits for this! I now plan to clean up / reorder the commits and then submit a proper pull request.

**Comment by n1tehawk** [comment]
Thinking about it, this change has the potential to annoy people as "being overly picky". Is there a way to keep the `6..` pattern disabled by default, or would using `--ignore` (or `--only`) be the solution?

**Comment by n1tehawk** [comment]
To preserve some of the commit message conversations:
>Ideally luacheck would also check for correct amount of whitespace around operators, parens, etc., but that would require more cooperation with the parser, and also it will be more complex to support customization. The warnings you suggested don't seem to need any options.

A proper lexer/parser-based implementation is surely a nice long-term goal. Though it would probably have to account for different coding styles / personal preferences.

For now, I've chosen a more "naive", but much simpler strategy - by using simple string patterns on the source lines. This is meant to detect common whitespace issues that easily get overlooked; and for which most coders will likely agree that they do constitute actual problems / should be addressed. Even this simple logic can easily point out many issues that aren't easily apparent (when visually inspecting the code), see e.g. 96d99de7 :wink:.

Admittedly this approach also has some shortcomings. A problem to be aware of is that is doesn't cope well with long string literals (`[[` and `]]` spanning multiple lines), where it will detect and report whitespace issues even if that might not be appropriate.

**Comment by mpeterv** [comment]
Barring the long string literals issue I don't think the is a reason to turn this off by default. I've added new warnings in many 0.x released and I think it's expected for a linter. Disabling the new warnings with '--ignore 6' if necessary is easy enough.
